platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/deprecation/DefaultDeprecationReporter.java

```diff
@@ -0,0 +1,83 @@
+import org.gradle.api.problems.deprecation.DeprecationReporter;
+import org.gradle.api.problems.deprecation.source.ReportSource;
+import org.gradle.api.problems.internal.DefaultProblemReporter;
+import org.gradle.api.problems.internal.GradleCoreProblemGroup;
+import org.gradle.api.problems.internal.InternalProblem;
+
+public class DefaultDeprecationReporter implements DeprecationReporter {
+
+    private final DefaultProblemReporter reporter;
+
+    public DefaultDeprecationReporter(DefaultProblemReporter reporter) {
+        this.reporter = reporter;
+    }
+
+    @Override
+    public void deprecate(ReportSource reportSource, String label, Action<DeprecateSpec> spec) {
+        DefaultDeprecationBuilder deprecationBuilder = new DefaultDeprecationBuilder(reportSource, reporter.createProblemBuilder());
+        deprecationBuilder.getProblemBuilder()
+            .id(GradleCoreProblemGroup.deprecation().generic())
+            .contextualLabel(label)
+            .stackLocation()
+            .withException(new RuntimeException());
+        spec.execute(deprecationBuilder);
+        reportBuiltProblem(deprecationBuilder);
+    }
+
+    @Override
+    public void deprecateMethod(ReportSource reportSource, Class<?> containingClass, String signature, Action<DeprecateMethodSpec> spec) {
+        DefaultDeprecationBuilder deprecationBuilder = new DefaultDeprecationBuilder(reportSource, reporter.createProblemBuilder());
+        String name = containingClass.getSimpleName() + "." + signature;
+        deprecationBuilder.getProblemBuilder()
+            .id(GradleCoreProblemGroup.deprecation().method())
+            .contextualLabel(String.format("Method '%s' is deprecated", name))
+            .stackLocation()
+            .withException(new RuntimeException());
+
+        spec.execute(deprecationBuilder);
+        reportBuiltProblem(deprecationBuilder);
+    }
+
+    @Override
+    public void deprecatePlugin(ReportSource reportSource, String pluginId, Action<DeprecatePluginSpec> spec) {
+        DefaultDeprecationBuilder deprecationBuilder = new DefaultDeprecationBuilder(reportSource, reporter.createProblemBuilder());
+        deprecationBuilder.getProblemBuilder()
+            .id(GradleCoreProblemGroup.deprecation().plugin())
+            .contextualLabel(String.format("Plugin '%s' is deprecated", pluginId))
+            .stackLocation()
+            .withException(new RuntimeException());
+        spec.execute(deprecationBuilder);
+        reportBuiltProblem(deprecationBuilder);
+    }
+
+    private Problem reportBuiltProblem(DefaultDeprecationBuilder builder) {
+        InternalProblem deprecationProblem = builder.build();
+        reporter.report(deprecationProblem);
+        return deprecationProblem;
+    }
+
+    // TODO (donat) cover indirectDeprecation: by default all deprecations are direct and so they should have stackLocation. The API should have an indirect() call on the spec to disable that.
```

Comment:
:x: remove todo. Consider adding an `indirect()` method to  `DeprecateSpec`.

Reply:
> Consider adding an indirect() method to DeprecateSpec.

I don't want to add it in this PR unless absolutely necessary.

Reply:
I would rather make an issue for this